snappy.cc

```diff
@@ -1102,14 +1158,9 @@ inline size_t AdvanceToNextTagARMOptimized(const uint8_t** ip_p, size_t* tag) {
   // instruction (csinc) and it removes several register moves.
   const size_t tag_type = *tag & 3;
   const bool is_literal = (tag_type == 0);
-  if (is_literal) {
-    size_t next_literal_tag = (*tag >> 2) + 1;
-    *tag = ip[next_literal_tag];
-    ip += next_literal_tag + 1;
-  } else {
-    *tag = ip[tag_type];
-    ip += tag_type + 1;
-  }
+  const size_t next_tag = is_literal ? (*tag >> 2) + 1 : tag_type;
+  *tag = ip[next_tag];
+  ip += (next_tag) + 1;
```

Comment:
This is the change that caused the regression on validate. I checked that with this change it is generating the desired `csinc` instruction, which the comment above says is the intended codegen. Here is the perf with just this change reverted:

```
BM_UFlat/0                       36099 ns        36098 ns        19334 bytes_per_second=2.64188G/s html
BM_UFlat/1                      457307 ns       457297 ns         1530 bytes_per_second=1.42986G/s urls
BM_UFlat/2                        3956 ns         3956 ns       176952 bytes_per_second=28.9819G/s jpg
BM_UFlat/3                         160 ns          160 ns      4394167 bytes_per_second=1.16619G/s jpg_200
BM_UFlat/4                        5788 ns         5788 ns       120704 bytes_per_second=16.478G/s pdf
BM_UFlat/5                      155126 ns       155115 ns         4509 bytes_per_second=2.45927G/s html4
BM_UFlat/6                      171845 ns       171839 ns         4073 bytes_per_second=844.067M/s txt1
BM_UFlat/7                      154670 ns       154670 ns         4525 bytes_per_second=771.836M/s txt2
BM_UFlat/8                      452391 ns       452362 ns         1547 bytes_per_second=899.687M/s txt3
BM_UFlat/9                      649864 ns       649864 ns         1077 bytes_per_second=707.13M/s txt4
BM_UFlat/10                      31264 ns        31263 ns        22349 bytes_per_second=3.53272G/s pb
BM_UFlat/11                     148201 ns       148197 ns         4721 bytes_per_second=1.15833G/s gaviota
BM_UFlatMedley                 2295899 ns      2295829 ns          305 bytes_per_second=1.188G/s
BM_UValidate/0                   20800 ns        20799 ns        33000 bytes_per_second=4.5851G/s html
BM_UValidate/1                  345447 ns       345427 ns         2020 bytes_per_second=1.89293G/s urls
BM_UValidate/2                     138 ns          138 ns      5059972 bytes_per_second=828.941G/s jpg
BM_UValidate/3                    96.5 ns         96.5 ns      7215460 bytes_per_second=1.93108G/s jpg_200
BM_UValidate/4                    1972 ns         1972 ns       354938 bytes_per_second=48.3615G/s pdf
BM_UValidate/5                  107941 ns       107936 ns         6443 bytes_per_second=3.53424G/s html4
BM_UValidate/6                  126895 ns       126888 ns         5501 bytes_per_second=1.11629G/s txt1
BM_UValidate/7                  114857 ns       114856 ns         6066 bytes_per_second=1039.39M/s txt2
BM_UValidate/8                  334870 ns       334859 ns         2087 bytes_per_second=1.18691G/s txt3
BM_UValidate/9                  486272 ns       486261 ns         1443 bytes_per_second=945.045M/s txt4
BM_UValidate/10                  17879 ns        17878 ns        38501 bytes_per_second=6.17749G/s pb
BM_UValidate/11                  92356 ns        92352 ns         7539 bytes_per_second=1.85876G/s gaviota
BM_UValidateMedley             1697835 ns      1697754 ns          412 bytes_per_second=1.6065G/s
BM_UIOVecSource/0               103097 ns       103092 ns         6599 bytes_per_second=947.274M/s html (22.24 %)
BM_UIOVecSource/1              1363926 ns      1363884 ns          512 bytes_per_second=490.923M/s urls (47.84 %)
BM_UIOVecSource/2                11296 ns        11296 ns        61854 bytes_per_second=10.1487G/s jpg (99.95 %)
BM_UIOVecSource/3                  402 ns          402 ns      1740237 bytes_per_second=475.039M/s jpg_200 (73.00 %)
BM_UIOVecSource/4                15463 ns        15463 ns        44743 bytes_per_second=6.16753G/s pdf (83.31 %)
BM_UIOVecSource/5               442147 ns       442130 ns         1580 bytes_per_second=883.507M/s html4 (22.52 %)
BM_UIOVecSource/6               416784 ns       416773 ns         1672 bytes_per_second=348.015M/s txt1 (57.87 %)
BM_UIOVecSource/7               367446 ns       367440 ns         1897 bytes_per_second=324.897M/s txt2 (62.02 %)
BM_UIOVecSource/8              1121587 ns      1121523 ns          622 bytes_per_second=362.885M/s txt3 (55.17 %)
BM_UIOVecSource/9              1522044 ns      1521967 ns          459 bytes_per_second=301.937M/s txt4 (66.41 %)
BM_UIOVecSource/10               90576 ns        90574 ns         7349 bytes_per_second=1.21938G/s pb (19.61 %)
BM_UIOVecSource/11              354641 ns       354634 ns         1954 bytes_per_second=495.669M/s gaviota (37.73 %)
BM_UIOVecSink/0                 102030 ns       102026 ns         6793 bytes_per_second=957.168M/s html
BM_UIOVecSink/1                 954178 ns       954179 ns          732 bytes_per_second=701.715M/s urls
BM_UIOVecSink/2                   4264 ns         4263 ns       164116 bytes_per_second=26.8888G/s jpg
BM_UIOVecSink/3                    321 ns          321 ns      2178205 bytes_per_second=593.638M/s jpg_200
BM_UIOVecSink/4                   9772 ns         9772 ns        70791 bytes_per_second=9.75947G/s pdf
BM_UFlatSink/0                   36097 ns        36096 ns        19360 bytes_per_second=2.64205G/s html
BM_UFlatSink/1                  457191 ns       457142 ns         1530 bytes_per_second=1.43034G/s urls
BM_UFlatSink/2                    3920 ns         3919 ns       178767 bytes_per_second=29.2498G/s jpg
BM_UFlatSink/3                     165 ns          165 ns      4253941 bytes_per_second=1.13175G/s jpg_200
BM_UFlatSink/4                    5820 ns         5820 ns       119850 bytes_per_second=16.3866G/s pdf
BM_UFlatSink/5                  154998 ns       154992 ns         4515 bytes_per_second=2.46122G/s html4
BM_UFlatSink/6                  171897 ns       171891 ns         4075 bytes_per_second=843.809M/s txt1
BM_UFlatSink/7                  154886 ns       154878 ns         4513 bytes_per_second=770.801M/s txt2
BM_UFlatSink/8                  452911 ns       452904 ns         1545 bytes_per_second=898.611M/s txt3
BM_UFlatSink/9                  649481 ns       649457 ns         1075 bytes_per_second=707.573M/s txt4
BM_UFlatSink/10                  31247 ns        31246 ns        22371 bytes_per_second=3.53464G/s pb
BM_UFlatSink/11                 148706 ns       148703 ns         4701 bytes_per_second=1.15439G/s gaviota
BM_ZFlat/0                       96115 ns        96112 ns         7178 bytes_per_second=1016.07M/s html (22.24 %)
BM_ZFlat/1                     1339282 ns      1339210 ns          521 bytes_per_second=499.968M/s urls (47.84 %)
BM_ZFlat/2                        6655 ns         6654 ns       104791 bytes_per_second=17.2275G/s jpg (99.95 %)
BM_ZFlat/3                         328 ns          328 ns      2136911 bytes_per_second=580.819M/s jpg_200 (73.00 %)
BM_ZFlat/4                       11728 ns        11728 ns        58781 bytes_per_second=8.13163G/s pdf (83.31 %)
BM_ZFlat/5                      424216 ns       424212 ns         1653 bytes_per_second=920.824M/s html4 (22.52 %)
BM_ZFlat/6                      411154 ns       411134 ns         1696 bytes_per_second=352.789M/s txt1 (57.87 %)
BM_ZFlat/7                      362120 ns       362110 ns         1925 bytes_per_second=329.679M/s txt2 (62.02 %)
BM_ZFlat/8                     1106496 ns      1106418 ns          630 bytes_per_second=367.84M/s txt3 (55.17 %)
BM_ZFlat/9                     1501500 ns      1501436 ns          465 bytes_per_second=306.066M/s txt4 (66.41 %)
BM_ZFlat/10                      86265 ns        86262 ns         7658 bytes_per_second=1.28032G/s pb (19.61 %)
BM_ZFlat/11                     347773 ns       347764 ns         1998 bytes_per_second=505.462M/s gaviota (37.73 %)
BM_ZFlatAll                    5898864 ns      5898386 ns          117 bytes_per_second=473.503M/s 12 kTestDataFiles
BM_ZFlatIncreasingTableSize      33035 ns        33035 ns        20928 bytes_per_second=938.586M/s 7 tables
```

It is a mixed bag for Validate, but even in the cases where it regresses there, it is still faster for actual decompression.